include/deal.II/base/mpi.h

```diff
@@ -500,6 +501,51 @@ namespace Utilities
                        const ArrayView<const T> &values,
                        const MPI_Comm           &mpi_communicator,
                        const ArrayView<T>       &output);
+      /**
+       * This function is intented for internal use in the library and
+       * aims at simplifying a many to many communication.
+       *
+       * @param[in] comm MPI communicator.
+       * @param[in] objects_to_send A map from the rank (unsigned int)
+       *  of the process meant to receive the data and the vector of
+       *  objects to send.
+       *
+       * @param[out] A map from the rank (unsigned int) of the process
+       *  which sent the data and the vector of objects received.
+       *
+       * Notice this is not just a point to point communication as the
+       * function uses an all-to-all communication to understand how
+       * many objects each process shall send, which is much slower.
+       *
+       * @author Giovanni Alzetta, Luca Heltai, 2017
+       */
+      template<typename T>
+      std::map<unsigned int, std::vector<T> >
+      send_and_receive(const MPI_Comm                       &comm,
+                       const std::map
+                       <unsigned int, typename std::vector<T> >
```

Comment:
technically, if you use boost serialization, you don't need a vector of objects, and could use an arbitrary type `T` here. As long as the underlying class is serializable this should work just fine for `std::vector<T>` as well.

I see that below you actually assume that `T` is a type (which is not clear from documentation).

Let's see what others have to say, but I would rework this function to use boost to serialize and drop `std::vector<>` altogether.

Reply:
You may want to look at how the new `GridTools::exchange_cell_data_to_ghosts()` function does the exchange, for example.

Reply:
BTW, you don't need the `typename` in this line.

Reply:
> You may want to look at how the new GridTools::exchange_cell_data_to_ghosts() function does the exchange, for example.

indeed.